homeassistant/components/songpal/config_flow.py

```diff
@@ -0,0 +1,158 @@
+        """Handle a flow start."""
+        # Check if already configured
+        if self._endpoint_already_configured():
+            return self.async_abort(reason="already_configured")
+
+        await self.async_set_unique_id(self.conf.endpoint)
+        self._abort_if_unique_id_configured()
+
+        if user_input is None:
+            return self.async_show_form(
+                step_id="init",
+                description_placeholders={
+                    CONF_NAME: self.conf.name,
+                    CONF_HOST: self.conf.host,
+                },
+            )
+
+        return self.async_create_entry(
+            title=self.conf.name,
+            data={
+                CONF_NAME: self.conf.name,
+                CONF_ENDPOINT: self.conf.endpoint,
+                CONF_MODEL: self.conf.model,
+            },
+        )
+
+    async def async_step_ssdp(self, discovery_info):
+        """Handle a discovered Songpal device."""
+        await self.async_set_unique_id(discovery_info[ssdp.ATTR_UPNP_UDN])
+        self._abort_if_unique_id_configured()
+
+        friendly_name = discovery_info[ssdp.ATTR_UPNP_FRIENDLY_NAME]
+        parsed_url = urlparse(discovery_info[ssdp.ATTR_SSDP_LOCATION])
+        model = discovery_info[ssdp.ATTR_UPNP_MODEL_NAME]
+        endpoint = discovery_info["X_ScalarWebAPI_DeviceInfo"]["X_ScalarWebAPI_BaseURL"]
+
+        # pylint: disable=no-member
+        self.context["title_placeholders"] = {
+            CONF_NAME: friendly_name,
+            CONF_HOST: parsed_url.hostname,
+        }
+
+        self.conf = SongpalConfig(friendly_name, parsed_url.hostname, endpoint, model,)
+
+        return await self.async_step_init()
+
+    async def async_step_import(self, user_input=None):
+        """Import a config entry."""
+        name = user_input.get(CONF_NAME)
+        endpoint = user_input.get(CONF_ENDPOINT)
+        parsed_url = urlparse(endpoint)
+        self.conf = SongpalConfig(name, parsed_url.hostname, endpoint)
+
+        errors = await self._async_try_connect()
+        if errors is not None:
+            _LOGGER.error(
+                "Unable to import songpal configuration (%s: %s). Please check your endpoint.",
+                name,
+                endpoint,
+            )
+            return self.async_abort(reason="connection")
+
+        return await self.async_step_init(user_input)
+
+    async def _async_try_connect(self):
+        """Try to connect and return errors."""
+        try:
+            device = Device(self.conf.endpoint)
+            await device.get_supported_methods()
+        except Exception as e:
+            _LOGGER.debug("Connection failed: %s", e)
+            return {"base": "connection"}
+        return None
+
+    def _endpoint_already_configured(self):
+        """See if we already have an endpoint matching user input configured."""
+        existing_endpoints = {
+            entry.data[CONF_ENDPOINT] for entry in self._async_current_entries()
```

Comment:
I see `_async_current_entries()` nowhere in the diff? If this is coming from the base class, is it okay to use a protected method?

If this is also a coroutine as the name indicates, it needs to be awaited.

Reply:
It comes from the base class. I'm not 100% sure if it is the right way to do this, but it is used in the config flow of many other integrations like [synology_dsm](https://github.com/home-assistant/core/blob/de3f0c788c04abf4b6c3039a8f2bb4d8f17d7344/homeassistant/components/synology_dsm/config_flow.py#L211) and [hue](https://github.com/home-assistant/core/blob/de3f0c788c04abf4b6c3039a8f2bb4d8f17d7344/homeassistant/components/hue/config_flow.py#L208) so I think it might be ok.

Reply:
Fair enough. Maybe that should not be marked as a private method, but it's not an issue of this PR.